From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Mon, 13 Jul 2026 00:57:46 +0800
Subject: [PATCH 05/47] smp: Do not check for CPU actually stopped on A7-A11.

Accessing cpu-impl-reg + 0x100 SErrors even when the core is powered on.
It seems the register does not exist on A7 - A11.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

---

TODO: check state using PMGR power state instead
---
 src/chickens.c | 4 ++++
 src/smp.c      | 2 +-
 src/utils.h    | 1 +
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/chickens.c b/src/chickens.c
index 128d3ed5..0c3b83ad 100644
--- a/src/chickens.c
+++ b/src/chickens.c
@@ -77,6 +77,7 @@ const struct midr_part_features features_m1 = {
     .mmu_sprr = true,
     .siq_cfg = true,
     .amx = true,
+    .impl_status = true,
 };
 
 const struct midr_part_features features_m2 = {
@@ -91,6 +92,7 @@ const struct midr_part_features features_m2 = {
     .siq_cfg = true,
     .amx = true,
     .actlr_el2 = true,
+    .impl_status = true,
 };
 
 const struct midr_part_features features_m3 = {
@@ -106,6 +108,7 @@ const struct midr_part_features features_m3 = {
     .amx = true,
     .actlr_el2 = true,
     .counter_redirect = true,
+    .impl_status = true,
 };
 
 // XXX figure out what features are actually available on M4
@@ -113,6 +116,7 @@ const struct midr_part_features features_m4 = {
     .sleep_mode = SLEEP_NONE, // XXX probably new mode required
     .fast_ipi = true,
     .actlr_el2 = true,
+    .impl_status = true,
 };
 
 /*
diff --git a/src/smp.c b/src/smp.c
index c9e97899..83eadff4 100644
--- a/src/smp.c
+++ b/src/smp.c
@@ -203,7 +203,7 @@ static void smp_stop_cpu(int index, int die, int cluster, int core, u64 impl, u6
 
     // If going into deep sleep, powering off the last core in a cluster kills our register
     // access, so just wait a bit.
-    if (deep_sleep) {
+    if (deep_sleep || !cpu_features->impl_status) {
         udelay(10000);
         printf("  Presumed stopped.\n");
         memset(&spin_table[index], 0, sizeof(struct spin_table));
diff --git a/src/utils.h b/src/utils.h
index 070cb29a..d675634a 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -510,6 +510,7 @@ struct midr_part_features {
     bool amx;
     bool actlr_el2;
     bool counter_redirect;
+    bool impl_status;
 };
 
 extern bool is_mac;
